### ISSUE TITLE
ngRepeat - Keep array in sync

### ISSUE THREAD

**Issue by jameswyse** [luckylooke/dragular#8]
Is it possible to automatically update the original array used by `ng-repeat` when moving items?

**Comment by luckylooke** [comment]
Hmm.. I will test it soon ;)

**Comment by jameswyse** [comment]
Agreed! Let me know if you need any help, I need this for a project with nested ngRepeats :)

**Comment by luckylooke** [comment]
I am thinking about three possible implementations:
- outside of dragular, just prepare dragular for supporting it
- inside dragular by getting info from ngRepeat, but ngRepeat is not revealing anything helpful, everything is keeped inside it. So we have to all that info harvest again (original collection array, filtered/manipulated array, item object, etc..)
- make own ngRepeat directive (ngRepeatDragular) and reveal all for dragularService

What do you think? :)

**Comment by rbatllet** [comment]
does this enhancement include dragging elements from an array scope to another array scope?

**Comment by jameswyse** [comment]
Parsing the ngRepeat expression or duplicating ngRepeat functionality will probably lead to problems down the road.

This might be too simplistic, but how about simply passing the array to the `dragularService` with the assumption that the indexes line up?

**Comment by luckylooke** [comment]
@rbatllet yes :)

@jameswyse thats good idea. Would you like to implement it?

**Comment by jameswyse** [comment]
Sure, though it may be a while before I have the time

**Comment by luckylooke** [comment]
no problem ;)

**Comment by luckylooke** [comment]
@jameswyse when you start working on it let me know here please, because if I will have time first, I`ll do it and let you know here. I just want to avoid the situation that both of us will be doindg the same think ok? :)

**Comment by jameswyse** [comment]
will do, thanks!

I'm thinking that it'd be easier to manipulate the data and let `ngRepeat` take care of the DOM, rather than trying to keep the data in sync.. but that may be a substantial change..

What do you think? Is there still a use-case for DOM-only reordering without a data source?

**Comment by luckylooke** [comment]
Hard to say, but in angular, I think that in everyone want to keep model
synced with dom in all cases. I cannot imagine the case without syncing.

**Comment by jameswyse** [comment]
Yeah that's what I figured :) 

I've kinda got it working in my project by using the events, but it's an ugly hack for sure.. I found that `ngRepeat` really doesn't like losing control of the DOM.. especially when the HTML comments (it uses them as markers) aren't moved with the element! It gets even trickier when you have multiple or nested repeats too :/ 

Basically what i'm doing at the moment is:

on the drag event
- store the original index somewhere so we can remove it from the array when dropped

on the drop event:
- try to find the container array by comparing the DOM indexes
- store the index of the new element inserted by dragular
- remove the element inserted by Dragular from the DOM
- splice the source array to remove the old item
- splice the destination array to insert the item at the correct index
- force ngRepeat to fully re-render - otherwise it goes mental on the next $digest (due to the HTML comments being messed up) 

I'm rushing to deliver this project but I should have some time over the weekend to look at this (if you/someone doesn't get to it first) :)

**Comment by luckylooke** [comment]
Ok.. if you manage it by the weekend thats fine. If not I will look at it
next week ;-)

**Comment by luckylooke** [comment]
Have you any progress in this feature? It is quite important feature, so I can implement it if you havent time for this. Just let me know :)

**Comment by luckylooke** [comment]
Since I started to need this and you are probably away, I begin to working on it.

**Comment by luckylooke** [comment]
I have make it work, I have it on my local branch, but I need to refactor examples site to test all features on it. So be patient..

**Comment by jameswyse** [comment]
Hey @luckylooke sorry, I was called away for work. Glad to see you've got it working though! Which way did you end up doing it?

**Comment by luckylooke** [comment]
Providing model via options.containersModel in same structure/order as
provided containers. :)

**Comment by jameswyse** [comment]
Nice!  Does it avoid the DOM manipulation entirely?

I'll have some time tonight and tomorrow, happy to help update the examples etc!

**Comment by luckylooke** [comment]
Yes in case you use models it does not manipulate dom directly and rather
updates models which cause dom update via scope.apply. I have splitted
exaples to subfolders for better maintenance. I need to finish it and make
model versions of each example to test all features. Im on vacation now but
in the evening I could have time for it.

**Comment by jameswyse** [comment]
Sounds great! Let me know if you push the code somewhere, I can help with the examples and testing :)

**Comment by luckylooke** [comment]
It is available in dev branch, but only first basic example is made also as model synced. There is some bug, that sometimes item is left in previous container, I cannot find out what circumstances are in such case. I think it can be caused by flickering during drag, maybe angular doesn't keep up with such fast changes. I will try to use some debounce techniques to avoid that.

**Comment by jameswyse** [comment]
Looking good!

I see what you mean about the bug.. It might help if we wait until the dragend event before updating the model? We'd have to move the placeholder/mirror element manually but that should be safe to do outside the $digest.

What do you think? I think I'll have a play with it and see what happens.

**Comment by luckylooke** [comment]
Ok, sounds good.. let me know when you get something ;)

**Comment by luckylooke** [comment]
I have fixed that bug and also flickering was removed... you can check it in dev branch now. I am gonna create and test it on other examples too. :)

**Comment by jameswyse** [comment]
:+1: I did something similar but was seeing some strange positioning bugs which seem to be happening in your demo too :/

Also we can't use `angular.element(el).scope()` as it's only available in debug mode

**Comment by luckylooke** [comment]
Can you give me more info about positioning bugs, I didn't notice them... and do you have idea how to bypass ```angular.element(el).scope()``` ? I didn't know it is only in debug, but I have checked it now in angular docs.. and you're right :/

**Comment by jameswyse** [comment]
In your demo, slowly drag item 5 downwards, you'll see that the mirror skips a slot and ends up after item 7.. Possibly a timing issue.

`scope.$apply()` actually triggers a $digest on the $rootScope, so you can call it on any scope you have access to. You could call it on the scope option used for events, or an injected `$rootScope`, or it might be better for performance to use `$timeout(fn, 0, false)`

It's a bit tricky because services shouldn't be manipulating the DOM at all, that's a job for directives :)

**Comment by jameswyse** [comment]
Ah I think I see the problem:

`_targetModel.splice(referenceIndex, 0, _lastTargetModel.splice(_currentIndex, 1)[0]);`

If you remove the dragged item from the array and it was originally before the reference item, then the `referenceIndex` will be wrong (in this case it's reduced by 1) - you could check for that or it might be better to recalculate `referenceIndex`?

**Comment by luckylooke** [comment]
Aha.. I will fix it, thanks.. and I will apply $timeout version of scope.$apply, thanks ;)

"services shouldn't be manipulating the DOM at all" - I know that, but isn't it cool as service? :D

**Comment by luckylooke** [comment]
I have fixed skipping a slot.
The code has still some problems (ngRepeat:dupes) with copy example... I hope I will fix it soon.. and then I will refactor scope.apply to $timeout.

**Comment by jameswyse** [comment]
Tried adding `track by $index` to the ngRepeat expression? :tractor:

**Comment by Astray-git** [comment]
A $timeout version of `scope.$apply`, is it something similar to [`scope.$applyAsync`](https://docs.angularjs.org/api/ng/type/$rootScope.Scope#$applyAsync)?

**Comment by luckylooke** [comment]
@jameswyse  should be working without 'track by'

**Comment by luckylooke** [comment]
@Astray-git I will try it, thanks :)

**Comment by luckylooke** [comment]
Bugs solved, works also without 'track by' as I originally thought it should be and scope.$apply replaced by $rootScope.$applyAsync. :)

**Comment by luckylooke** [comment]
Now I have problem with models in directive dragular. I dont think there is posibility to provide model via JSON ```dragular='{"containersModel": ?????? ,"nameSpace":"common"}'``. There is second option how to provide options to directive. Via scope-property name ```dragular="dragularOptions"```, so it seems to be only option in this case, or?

**Comment by luckylooke** [comment]
I would need to implement some angular expression support for directive to avoid this:
```html
<div class='container' dragular="dragularOptions1">
        <div ng-repeat="item in items1">{{item.content}}</div>
      </div>
      <div class='container' dragular='dragularOptions2'>
        <div ng-repeat="item in items2">{{item.content}}</div>
      </div>
```
```js
.controller('DirectiveModel', ['$scope', '$element', 'dragularService', function TodoCtrl($scope, $element, dragularService) {
    $scope.items1 = [{
      content: 'Item 1'
    }, {
      content: 'Item 2'
    }, {
      content: 'Item 3'
    }, {
      content: 'Item 4'
    }];
    $scope.items2 = [{
      content: 'Item 5'
    }, {
      content: 'Item 6'
    }, {
      content: 'Item 7'
    }, {
      content: 'Item 8'
    }];
    $scope.dragularOptions1 = {
      containersModel: [$scope.items1],
      classes: {
        mirror: 'custom-green-mirror'
      },
      nameSpace: 'common' // just connecting left and right container
    };
    $scope.dragularOptions2 = {
      containersModel: [$scope.items2],
      classes: {
        mirror: 'custom-green-mirror'
      },
      nameSpace: 'common' // just connecting left and right container
    };
  }])
```
Then I can apply same for JSON format:
```js
dragular='{"containersModel":"angular expression","nameSpace":"common"}'
```

**Comment by Astray-git** [comment]
I'm just wondering why we need to put containerModel in a JSON? I think in this case we provide the containerModel from a controller, then we can config with  `dragular="controller.dragularOptions"`

**Comment by luckylooke** [comment]
@Astray-git do you have link to some live example passing config from controller to directive like this ```dragular="controller.dragularOptions"```? Like jsfiddle, codepen, .. etc. Thanks

**Comment by luckylooke** [comment]
now option.containersModel supports expressions wueey.
```html
 <div class='container' dragular="dragularOptions">
        <div ng-repeat="item in items1">{{item.content}}</div>
      </div>
      <div class='container' dragular='{"containersModel":"[items2]","classes":{"mirror":"custom-green-mirror"},"nameSpace":"common"}'>
        <div ng-repeat="item in items2">{{item.content}}</div>
      </div>
```
```js
.controller('DirectiveModel', ['$scope', '$element', 'dragularService', function TodoCtrl($scope, $element, dragularService) {
    $scope.items1 = [{
      content: 'Move me, and make copy on drop.'
    }, {
      content: 'If you try to drop me somewhere other than these containers, I\'ll just come back.'
    }, {
      content: 'Item 3'
    }, {
      content: 'Item 4'
    }];
    $scope.items2 = [{
      content: 'Item 5'
    }, {
      content: 'Item 6'
    }, {
      content: 'Item 7'
    }, {
      content: 'Item 8'
    }];
    $scope.dragularOptions = {
      containersModel: [$scope.items1],
      classes: {
        mirror: 'custom-green-mirror'
      },
      nameSpace: 'common' // just connecting left and right container
    };
  }])
```

**Comment by jameswyse** [comment]
You could use the `ngModel` directive and pass options separately? It's commonly used in this situation :)

https://docs.angularjs.org/api/ng/directive/ngModel

**Comment by jameswyse** [comment]
great work by the way!

**Comment by Astray-git** [comment]
@luckylooke http://jsbin.com/yisavopeme/edit?html,js,output

**Comment by luckylooke** [comment]
@Astray-git thanks.. I will put it into docs later :+1:

**Comment by luckylooke** [comment]
@jameswyse I think there is plenty options now how to pass options into dragular, we will not need ngModel as another one :)

**Comment by luckylooke** [comment]
Today I am going to finish all examples and then I have a lot todos with code, after that I will finally release new version! :cake:

**Comment by luckylooke** [comment]
fixed and published in release 2.0.0
